### PR TITLE
enhancement on DB operations in import

### PR DESCRIPTION
**Pull request by immarvin**
Description:
* fix the non-group objects import issue https://github.com/xcat2/xcat-inventory/issues/85 , the multiple key table related issues https://github.com/xcat2/xcat-inventory/issues/140 and  https://github.com/xcat2/xcat-inventory/issues/138, the `import -c` support for site object issue https://github.com/xcat2/xcat-inventory/issues/139
* support the mapping from multiple key table to individual objects or subset attributes of an object
* support the reference of sub inventory objects in object definition yaml with `REF{<sub inventory object>}` syntax
* the inventory schema grows to version `2.0`, a sub inventory type `networkconn` is added to describe the network connection info in `switch` table
UT:

1. switch table is mapped to `network_info. connections` of node object , export node object:
```
[root@c910f03c05k21 inventory]# xcat-inventory export -t node -o testnodes
node:
  testnodes:
    device_info:
      arch: armel
      characteristics: switch
      switchtype: onie
    device_type: switch
    engines:
      hardware_mgt_engine:
        engine_type: switch
      netboot_engine:
        engine_info:
          osimage: cumulus3.5.2-armel
        engine_type: onie
    network_info:
      connections:
      - switch: mid05tor11
        switchport: '22'
      - switch: mid05tor13
        switchport: '22'
      - switch: mid05tor14
        switchport: '22'
      primarynic:
        ip: 10.3.5.141
        mac:
        - 8c:ea:2b:12:ca:40
    obj_info:
      description: Cumulus Linux 3.4.1 (Linux Kernel 4.1.33-1+cl3u9)
      groups: switch,edge_switch
    obj_type: node
    role: compute
    role_info:
      tftpserver: briggs01
schema_version: '2.0'

#Version 2.14.5 (git commit 5bab3f8bac3374cd9cb2e26b4a71c7cb32bb1126, built Mon Oct 22 06:16:06 EDT 2018)
[root@c910f03c05k21 inventory]#

```

2. import node object:
```
[root@c910f03c05k21 inventory]# xcat-inventory import -f /tmp/testnodes
loading inventory date in "/tmp/testnodes"
start to import "node" type objects
 preprocessing "node" type objects
 writting "node" type objects
start to import "networkconn" type objects
 preprocessing "networkconn" type objects
 writting "networkconn" type objects
Inventory import successfully!
[root@c910f03c05k21 inventory]# xcat-inventory export -t node -o testnodes
node:
  testnodes:
    device_info:
      arch: armel
      characteristics: switch
      switchtype: onie
    device_type: switch
    engines:
      hardware_mgt_engine:
        engine_type: switch
      netboot_engine:
        engine_info:
          osimage: cumulus3.5.2-armel
        engine_type: onie
    network_info:
      connections:
      - switch: mid05tor11
        switchport: '22'
      - switch: mid05tor13
        switchport: '22'
      - switch: mid05tor14
        switchport: '24'
      primarynic:
        ip: 10.3.5.141
        mac:
        - 8c:ea:2b:12:ca:40
    obj_info:
      description: Cumulus Linux 3.4.1 (Linux Kernel 4.1.33-1+cl3u9)
      groups: switch,edge_switch
    obj_type: node
    role: compute
    role_info:
      tftpserver: briggs01
schema_version: '2.0'
```

3. export of `passwd`:
```
[root@c910f03c05k21 inventory]# xcat-inventory export -t passwd
passwd:
  ipmi:
    password: '''USERID'''
    username: '''admin'''
  openbmc:
  - password: xcat
    username: mmmmmmmm
  - password: USERID
    username: ssss
  switch:
    password: admin
    username: root
  system:
  - password: cluster
    username: root
  - password: cluster
    username: wsuser
schema_version: '2.0'
```

4. import of `passwd`:
```
[root@c910f03c05k21 inventory]# cat /tmp/passwd
passwd:
  ipmi:
    password: '''USERID'''
    username: '''admin'''
  openbmc:
  - password: 'USERID'
    username: 'ssss'
  switch:
    password: admin
    username: root
  system:
  - password: cluster
    username: root
  - password: cluster
    username: wsuser
schema_version: '2.0'

#Version 2.14.5 (git commit 5bab3f8bac3374cd9cb2e26b4a71c7cb32bb1126, built Mon Oct 22 06:16:06 EDT 2018)
[root@c910f03c05k21 inventory]# xcat-inventory import -f /tmp/passwd -c
loading inventory date in "/tmp/passwd"
start to import "passwd" type objects
 preprocessing "passwd" type objects
 writting "passwd" type objects
Inventory import successfully!
[root@c910f03c05k21 inventory]# tabdump passwd
#key,username,password,cryptmethod,authdomain,comments,disable
"ipmi","'admin'","'USERID'",,,,
"openbmc","ssss","USERID",,,,
"switch","root","admin",,,,
"system","root","cluster",,,,
"system","wsuser","cluster",,,,
[root@c910f03c05k21 inventory]#
```

### CODE REVIEW

File: xcat-inventory/xcclient/inventory/schema/2.0/1 [hunk 1]
```diff
@@ -0,0 +1,280 @@
+!!python/dict
```
Comment (bybai) [inline xcat-inventory/xcclient/inventory/schema/2.0/1]: `xcat-inventory/xcclient/inventory/schema/2.0/1` is not valid schema file name.

File: xcat-inventory/xcclient/inventory/dbfactory.py [hunk 2]
```diff
@@ -24,18 +32,20 @@ def create_or_update(session,tabcls,key,newdict,ismatrixtable=True):
...
-        if  tabkey != item and newdict[item] is not None:
+        if item not in objkeys and newdict[item]:
             skiprow=0        
+            delrow=0
 
-        if  tabkey != item and newdict[item] is None:
+        if  item not in objkeys and newdict[item] is None:
             newdict[item]=''
         
         #delete table rows when (1)the key is None or blank (2)the key is not specified in newdict and all non-key values are blank
```
Comment (bybai) [inline xcat-inventory/xcclient/inventory/dbfactory.py]: could you update the comments according to new codes?

File: xcat-inventory/xcclient/inventory/schema/2.0/1 (hunk 1 above)
Comment (immarvin) [reply xcat-inventory/xcclient/inventory/schema/2.0/1]: fixed

File: xcat-inventory/xcclient/inventory/dbfactory.py [hunk 3]
```diff
@@ -46,77 +56,129 @@ def create_or_update(session,tabcls,key,newdict,ismatrixtable=True):
...
+        #print("=========matrixdbfactory:settab========")
         #print(tabdict)
         #print("\n")
         if tabdict is None:
             return None
         for key in tabdict.keys():
             utils.verbose("  writting object: "+str(key),file=sys.stdout)
+            #clear any existing table entries before adding new entries
+            df=dbfactory(self._dbsession) 
+            df.cleartab(tabdict[key].keys(),[key])
```
Comment (bybai) [inline xcat-inventory/xcclient/inventory/dbfactory.py]: When it is not update, `manager.py` calls `cleartab` to clear up db, there is no data in DB now. In line 171 call cleartab, it clear up DB again in for loop. And if there are too many nodes, for each node, it  clear up `switch` table once, the performance can be enhanced here.

File: xcat-inventory/xcclient/inventory/dbfactory.py (hunk 2 above)
Comment (immarvin) [reply xcat-inventory/xcclient/inventory/dbfactory.py]: fixed

File: xcat-inventory/xcclient/inventory/dbfactory.py (hunk 3 above)
Comment (immarvin) [reply xcat-inventory/xcclient/inventory/dbfactory.py]: will refine this in following sprints

File: xcat-inventory/xcclient/inventory/dbfactory.py [hunk 4]
```diff
@@ -244,23 +331,27 @@ def cleartab(self,tabs):
                 flattabs.append(tab)
             else:
                 matrixtabs.append(tab)
-        for tab in matrixtabs:
+        for tab in tabs:
             if hasattr(dbobject,tab):
                 tabcls=getattr(dbobject,tab)
```
[the remark this answers is not in the document]
Comment (immarvin) [reply xcat-inventory/xcclient/inventory/dbfactory.py]: fixed

File: xcat-inventory/xcclient/inventory/dbfactory.py (hunk 3 above)
Comment (bybai) [reply xcat-inventory/xcclient/inventory/dbfactory.py]: I opened a ticket to trace this enhance https://github.com/xcat2/xcat-inventory/issues/147, thanks

**Comment by bybai** [comment]
I agree to merge this pull request for functions fixing , the DB execution performance enhancement ticket (https://github.com/xcat2/xcat-inventory/issues/147) can be planned in another sprint. Thanks

**Pull request merged**